src/Features/Core/Portable/ExtractClass/ExtractClassWithDialogCodeAction.cs

```diff
@@ -0,0 +1,264 @@
+                        extractClassOptions.FileName,
+                        _document.Project.Id,
+                        _document.Folders,
+                        newType,
+                        fileBanner,
+                        cancellationToken).ConfigureAwait(false);
+
+                // Update the original type to have the new base
+                var solutionWithUpdatedOriginalType = await GetSolutionWithBaseAddedAsync(
+                    updatedDocument.Project.Solution,
+                    symbolMapping,
+                    newType,
+                    extractClassOptions.MemberAnalysisResults,
+                    cancellationToken).ConfigureAwait(false);
+
+                // After all the changes, make sure we're using the most up to date symbol 
+                // as the destination for pulling members into
+                var documentWithTypeDeclaration = solutionWithUpdatedOriginalType.GetRequiredDocument(updatedDocument.Id);
+                newType = await GetNewTypeSymbolAsync(documentWithTypeDeclaration, typeAnnotation, cancellationToken).ConfigureAwait(false);
+
+                // Use Members Puller to move the members to the new symbol
+                var finalSolution = await PullMembersUpAsync(
+                    solutionWithUpdatedOriginalType,
+                    newType,
+                    symbolMapping,
+                    extractClassOptions.MemberAnalysisResults,
+                    cancellationToken).ConfigureAwait(false);
+
+                return new[] { new ApplyChangesOperation(finalSolution) };
+            }
+            else
+            {
+                // If user click cancel button, options will be null and hit this branch
+                return SpecializedCollections.EmptyEnumerable<CodeActionOperation>();
+            }
+        }
+
+        private async Task<Solution> PullMembersUpAsync(
+            Solution solution,
+            INamedTypeSymbol newType,
+            AnnotatedSymbolMapping symbolMapping,
+            ImmutableArray<ExtractClassMemberAnalysisResult> memberAnalysisResults,
+            CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<(ISymbol member, bool makeAbstract)>.GetInstance(out var pullMembersBuilder);
+            using var _1 = ArrayBuilder<ExtractClassMemberAnalysisResult>.GetInstance(memberAnalysisResults.Length, out var remainingResults);
+            remainingResults.AddRange(memberAnalysisResults);
+
+            // For each document in the symbol mappings, we want to find the annotated nodes
+            // of the members and get the current symbol that represents those after
+            // any changes we made before members get pulled up into the base class.
+            // We only need to worry about symbols that are actually being moved, so we track
+            // the symbols from the member analysis. 
+            foreach (var documentId in symbolMapping.DocumentIds)
```

Comment:
feels like AnnotatedSymbolMapping should group by DocumentId to results in taht doc.  that way you can process a doc at a time. t his approach is weird where you just walk the docs, and then lookup things randomly in them, just expecting they may fail.